File: app/src/main/java/jp/juggler/subwaytooter/table/PostDraft.java
package jp.juggler.subwaytooter.table;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.provider.BaseColumns;
import android.support.annotation.NonNull;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.util.LogCategory;
import jp.juggler.subwaytooter.util.Utils;

public class PostDraft {
	
	private static final LogCategory log = new LogCategory( "PostDraft" );
	
	private static final String table = "post_draft";
	private static final String COL_ID = BaseColumns._ID;
	private static final String COL_TIME_SAVE = "time_save";
	private static final String COL_JSON = "json";
	private static final String COL_HASH = "hash";
	
	public static void onDBCreate( SQLiteDatabase db ){
		log.d( "onDBCreate!" );
		db.execSQL(
			"create table if not exists " + table
				+ "(" + COL_ID + " INTEGER PRIMARY KEY"
				+ "," + COL_TIME_SAVE + " integer not null"
				+ "," + COL_JSON + " text not null"
				+ "," + COL_HASH + " text not null"
				+ ")"
		);
		db.execSQL(
			"create unique index if not exists " + table + "_hash on " + table + "(" + COL_HASH + ")"
		);
		db.execSQL(
			"create index if not exists " + table + "_time on " + table + "(" + COL_TIME_SAVE + ")"
		);
	}
	
	public static void onDBUpgrade( SQLiteDatabase db, int oldVersion, int newVersion ){
		if( oldVersion < 12 && newVersion >= 12 ){
			onDBCreate( db );
		}
	}
	
	private static void deleteOld( long now ){
		try{
			// 古いデータを掃除する
			long expire = now - 86400000L * 30;
			App1.getDB().delete( table, COL_TIME_SAVE + "<?", new String[]{ Long.toString( expire ) } );
		}catch( Throwable ex ){
			log.e( ex, "deleteOld failed." );
		}
	}
	
	public static void save( long now, @NonNull JSONObject json ){
		
		deleteOld( now );
		
		try{
			// make hash
			StringBuilder sb = new StringBuilder();
			ArrayList< String > keys = new ArrayList<>();
			for( Iterator< String > it = json.keys() ; it.hasNext() ; ){
				keys.add( it.next() );
			}
			Collections.sort( keys );
			for( String k : keys ){
				String v = json.isNull( k ) ? "(null)" : json.opt( k ).toString();
				sb.append( "&" );
				sb.append( k );
				sb.append( "=" );
				sb.append( v );
			}
			String hash = Utils.digestSHA256( sb.toString() );
			
			// save to db
			ContentValues cv = new ContentValues();
			cv.put( COL_TIME_SAVE, now );
			cv.put( COL_JSON, json.toString() );
			cv.put( COL_HASH, hash );
			App1.getDB().replace( table, null, cv );
		}catch( Throwable ex ){
			log.e( ex, "save failed." );
		}
	}

	public static boolean hasDraft(){
		try{
			Cursor cursor = App1.getDB().query( table, new String[]{ "count(*)" }, null, null, null, null, null );
			if( cursor != null ){
				try{
					if( cursor.moveToNext() ){
						int count = cursor.getInt( 0 );
						return count > 0;
					}
				}finally{
					cursor.close();
				}
			}
		}catch( Throwable ex ){
			log.trace( ex );
			log.e( ex, "hasDraft failed." );
		}
		return false;
	}
	
	public static Cursor createCursor(){
		try{
			return App1.getDB().query( table, null, null, null, null, null, COL_TIME_SAVE + " desc" );
		}catch( Throwable ex ){
			log.trace( ex );
			log.e( ex, "createCursor failed." );
		}
		return null;
	}
	
	public static class ColIdx {
		final int idx_id;
		final int idx_time_save;
		final int idx_json;
		final int idx_hash;
		
		public ColIdx( Cursor cursor ){
			idx_id = cursor.getColumnIndex( COL_ID );
			idx_time_save = cursor.getColumnIndex( COL_TIME_SAVE );
			idx_json = cursor.getColumnIndex( COL_JSON );
			idx_hash = cursor.getColumnIndex( COL_HASH );
		}
		
	}
	
	@SuppressWarnings("WeakerAccess") public long id;
	@SuppressWarnings("WeakerAccess") public long time_save;
	@SuppressWarnings("WeakerAccess") public JSONObject json;
	@SuppressWarnings("WeakerAccess") public String hash;
	
	public static PostDraft loadFromCursor( Cursor cursor, ColIdx colIdx, int position ){
		if( colIdx == null ) colIdx = new ColIdx( cursor );

		if( ! cursor.moveToPosition( position ) ){
			log.d("loadFromCursor: move failed. position=%s",position);
			return null;
		}

		PostDraft dst = new PostDraft();
		dst.id = cursor.getLong( colIdx.idx_id );
		dst.time_save = cursor.getLong( colIdx.idx_time_save );
		try{
			dst.json = new JSONObject( cursor.getString( colIdx.idx_json ) );
		}catch( Throwable ex ){
			log.trace( ex );
			dst.json = new JSONObject();
		}
		dst.hash = cursor.getString( colIdx.idx_hash );
		return dst;
	}
	
	public void delete(){
		try{
			App1.getDB().delete( table, COL_ID + "=?", new String[]{ Long.toString( id ) } );
		}catch( Throwable ex ){
			log.e( ex, "delete failed." );
		}
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/table/PostDraft.kt
package jp.juggler.subwaytooter.table

import android.annotation.SuppressLint
import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import jp.juggler.util.*
import jp.juggler.util.data.*
import jp.juggler.util.log.LogCategory

class PostDraft(
    var id: Long = 0,
    var time_save: Long = 0,
    var json: JsonObject? = null,
    var hash: String? = null,
) {

    companion object : TableCompanion {

        private val log = LogCategory("PostDraft")

        override val table = "post_draft"
        private const val COL_ID = BaseColumns._ID
        private const val COL_TIME_SAVE = "time_save"
        private const val COL_JSON = "json"
        private const val COL_HASH = "hash"

        override fun onDBCreate(db: SQLiteDatabase) {
            log.d("onDBCreate!")
            db.execSQL(
                """create table if not exists $table
                ($COL_ID INTEGER PRIMARY KEY
                ,$COL_TIME_SAVE integer not null
                ,$COL_JSON text not null
                ,$COL_HASH text not null
                )""".trimIndent()
            )
            db.execSQL("create unique index if not exists ${table}_hash on $table($COL_HASH)")
            db.execSQL("create index if not exists ${table}_time on $table($COL_TIME_SAVE)")
        }

        override fun onDBUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            if (oldVersion < 12 && newVersion >= 12) {
                onDBCreate(db)
            }
        }
    }

    class ColIdx(cursor: Cursor) {
        private val idxId = cursor.getColumnIndex(COL_ID)
        private val idxTimeSave = cursor.getColumnIndex(COL_TIME_SAVE)
        private val idxJson = cursor.getColumnIndex(COL_JSON)
        private val idxHash = cursor.getColumnIndex(COL_HASH)
        fun readRow(cursor: Cursor) = PostDraft(
            id = cursor.getLong(idxId),
            time_save = cursor.getLong(idxTimeSave),
            hash = cursor.getString(idxHash),
            json = try {
                cursor.getString(idxJson).decodeJsonObject()
            } catch (ex: Throwable) {
                log.e(ex, "loadFromCursor failed.")
                JsonObject()
            }
        )
    }

    class Access(val db: SQLiteDatabase) {

        private fun deleteOld(now: Long) {
            try {
                // 古いデータを掃除する
                val expire = now - 86400000L * 30
                db.delete(table, "$COL_TIME_SAVE<?", arrayOf(expire.toString()))
            } catch (ex: Throwable) {
                log.e(ex, "deleteOld failed.")
            }
        }

        fun save(now: Long, json: JsonObject) {
            deleteOld(now)
            try {
                // make hash
                val hash = StringBuilder().also { sb ->
                    json.keys.sorted().forEach { k ->
                        val v = json[k]?.toString() ?: "(null)"
                        sb.append("&")
                        sb.append(k)
                        sb.append("=")
                        sb.append(v)
                    }
                }.toString().digestSHA256Hex()

                // save to db
                db.replace(table, null, ContentValues().apply {
                    put(COL_TIME_SAVE, now)
                    put(COL_JSON, json.toString())
                    put(COL_HASH, hash)
                })
            } catch (ex: Throwable) {
                log.e(ex, "save failed.")
            }
        }

        fun delete(item: PostDraft) {
            try {
                db.deleteById(table, item.id.toString(), COL_ID)
            } catch (ex: Throwable) {
                log.e(ex, "delete failed.")
            }
        }

        fun hasDraft(): Boolean {
            try {
                db.query(table, arrayOf("count(*)"), null, null, null, null, null)
                    .use { cursor ->
                        if (cursor.moveToNext()) {
                            val count = cursor.getInt(0)
                            return count > 0
                        }
                    }
            } catch (ex: Throwable) {
                log.e(ex, "hasDraft failed.")
            }
            return false
        }

        // caller must close the cursor
        @SuppressLint("Recycle")
        fun createCursor(): Cursor =
            db.queryAll(table, "$COL_TIME_SAVE desc")!!

        fun loadFromCursor(
            cursor: Cursor,
            colIdxArg: ColIdx?,
            position: Int,
        ): PostDraft? = when {
            cursor.moveToPosition(position) ->
                (colIdxArg ?: ColIdx(cursor)).readRow(cursor)
            else -> {
                log.d("loadFromCursor: move failed. position=$position")
                null
            }
        }
    }
}
